File: dotnet-backend/Assetflow.NotificationService/NotificationDataService.cs
using System.Net;
using System.Net.Mail;
using System.Text.Json;
using MySqlConnector;

namespace Assetflow.NotificationService;

public sealed class NotificationDataService(IConfiguration configuration, IHttpClientFactory httpClientFactory)
{
    private readonly string _connectionString = configuration.GetConnectionString("DefaultConnection")
        ?? $"Server={configuration["Database:Host"] ?? "localhost"};Port={configuration["Database:Port"] ?? "3306"};Database={configuration["Database:Name"] ?? "assetflow_db"};User ID={configuration["Database:User"] ?? "root"};Password={configuration["Database:Password"] ?? Environment.GetEnvironmentVariable("SPRING_DATASOURCE_PASSWORD") ?? "Astrixz"};Allow User Variables=True;";

    private MySqlConnection Connection() => new(_connectionString);
    private static long Number(IReadOnlyDictionary<string, string?> claims, string name) => long.TryParse(claims.GetValueOrDefault(name), out var value) ? value : 0;
    private static long Company(IReadOnlyDictionary<string, string?> claims) => Number(claims, "companyId");
    private static long User(IReadOnlyDictionary<string, string?> claims) => Number(claims, "sub");
    private static object? Value(IReadOnlyDictionary<string, object?> body, string name, object? fallback = null) => Normalize(body.GetValueOrDefault(name) ?? fallback);
    private static object? Normalize(object? value) => value is not JsonElement json ? value : json.ValueKind switch
    {
        JsonValueKind.Null => null,
        JsonValueKind.String => json.GetString(),
        JsonValueKind.Number when json.TryGetInt64(out var integer) => integer,
        JsonValueKind.Number when json.TryGetDecimal(out var decimalValue) => decimalValue,
        JsonValueKind.True => true,
        JsonValueKind.False => false,
        _ => json.GetRawText()
    };

    public async Task<List<Dictionary<string, object?>>> NotificationsAsync(IReadOnlyDictionary<string, string?> c)
    {
        await using var db = Connection(); await db.OpenAsync();
        return await QueryAsync(db, "select * from notifications where company_id=@company and (user_id=@user or user_id=0) order by created_at desc", ("@company", Company(c)), ("@user", User(c)));
    }

    public async Task<Dictionary<string, object?>> CreateNotificationAsync(IReadOnlyDictionary<string, string?> c, IReadOnlyDictionary<string, object?> body)
    {
        await using var db = Connection(); await db.OpenAsync();
        await using var cmd = new MySqlCommand("insert into notifications(company_id,user_id,title,message) values(@company,@user,@title,@message)", db);
        cmd.Parameters.AddWithValue("@company", Company(c)); cmd.Parameters.AddWithValue("@user", Value(body, "userId", User(c)));
        cmd.Parameters.AddWithValue("@title", Value(body, "title", "AssetFlow notification")); cmd.Parameters.AddWithValue("@message", Value(body, "message", ""));
        await cmd.ExecuteNonQueryAsync(); return body.ToDictionary(pair => pair.Key, pair => pair.Value);
    }

    public Task<int> ReadAsync(IReadOnlyDictionary<string, string?> c, long id) => ExecuteAsync("update notifications set is_read=true where notification_id=@id and company_id=@company", ("@id", id), ("@company", Company(c)));
    public Task<int> DeleteAsync(IReadOnlyDictionary<string, string?> c, long id) => ExecuteAsync("delete from notifications where notification_id=@id and company_id=@company and (user_id=@user or user_id=0)", ("@id", id), ("@company", Company(c)), ("@user", User(c)));
    public Task<int> ReadAllAsync(IReadOnlyDictionary<string, string?> c) => ExecuteAsync("update notifications set is_read=true where company_id=@company and (user_id=@user or user_id=0)", ("@company", Company(c)), ("@user", User(c)));

    public async Task<List<Dictionary<string, object?>>> AuditsAsync(IReadOnlyDictionary<string, string?> c)
    {
        await using var db = Connection(); await db.OpenAsync();
        var employee = string.Equals(c.GetValueOrDefault("role"), "EMPLOYEE", StringComparison.Ordinal);
        var sql = employee ? "select * from audit_logs where company_id=@company and user_id=@user order by created_at desc limit 200" : "select * from audit_logs where company_id=@company order by created_at desc limit 200";
        return await QueryAsync(db, sql, ("@company", Company(c)), ("@user", User(c)));
    }

    public async Task<Dictionary<string, object?>> CreateAuditAsync(IReadOnlyDictionary<string, string?> c, IReadOnlyDictionary<string, object?> body)
    {
        await using var db = Connection(); await db.OpenAsync();
        await using var cmd = new MySqlCommand("insert into audit_logs(company_id,user_id,module,action,entity_id,description,ip_address) values(@company,@user,@module,@action,@entity,@description,@ip)", db);
        cmd.Parameters.AddWithValue("@company", Company(c)); cmd.Parameters.AddWithValue("@user", User(c)); cmd.Parameters.AddWithValue("@module", Value(body, "module", "unknown")); cmd.Parameters.AddWithValue("@action", Value(body, "action", "unknown")); cmd.Parameters.AddWithValue("@entity", Value(body, "entityId")); cmd.Parameters.AddWithValue("@description", Value(body, "description")); cmd.Parameters.AddWithValue("@ip", Value(body, "ipAddress"));
        await cmd.ExecuteNonQueryAsync(); return new(body);
    }

    public async Task<Dictionary<string, long>> DashboardAsync(IReadOnlyDictionary<string, string?> c)
    {
        var employee = string.Equals(c.GetValueOrDefault("role"), "EMPLOYEE", StringComparison.Ordinal); var scope = employee ? " and user_id=@user" : "";
        return new() { ["notifications"] = await CountAsync($"select count(*) from notifications where company_id=@company{scope}", c, employee), ["unreadNotifications"] = await CountAsync($"select count(*) from notifications where company_id=@company and is_read=false{scope}", c, employee), ["auditEvents"] = await CountAsync($"select count(*) from audit_logs where company_id=@company{scope}", c, employee) };
    }

    public async Task<List<Dictionary<string, object?>>> ReportAsync(string type, IReadOnlyDictionary<string, string?> c)
    {
        var sql = type.ToLowerInvariant() switch { "assets" => "select status,count(*) total from assets where company_id=@company group by status", "users" or "employees" => "select r.role_name,count(*) total from users u join roles r on r.role_id=u.role_id where u.company_id=@company group by r.role_name", "departments" => "select d.department_name,count(u.user_id) total from departments d left join users u on u.department_id=d.department_id and u.company_id=d.company_id where d.company_id=@company group by d.department_id,d.department_name", "vendors" => "select is_active,count(*) total from vendors where company_id=@company group by is_active", "maintenance" => "select status,count(*) total from service_tickets where company_id=@company group by status", "requests" => "select status,count(*) total from asset_requests where company_id=@company group by status", "audit" => "select module,count(*) total from audit_logs where company_id=@company group by module order by total desc", _ => "select status,count(*) total from purchase_orders where company_id=@company group by status" };
        await using var db = Connection(); await db.OpenAsync(); return await QueryAsync(db, sql, ("@company", Company(c)));
    }

    public async Task<Dictionary<string, object?>> ChatAsync(IReadOnlyDictionary<string, object?> body, IReadOnlyDictionary<string, string?>? claims = null)
    {
        var key = configuration["Gemini:ApiKey"] ?? Environment.GetEnvironmentVariable("GEMINI_API_KEY"); var question = Convert.ToString(Value(body, "message", "")) ?? "";
        if (string.IsNullOrWhiteSpace(key)) return new() { ["reply"] = "Gemini is not configured yet. Set GEMINI_API_KEY to enable company-aware answers.", ["configured"] = false };
        var model = configuration["Gemini:Model"] ?? Environment.GetEnvironmentVariable("GEMINI_MODEL") ?? "gemini-2.0-flash";
        var payload = new { contents = new[] { new { parts = new[] { new { text = question } } } } };
        try { var client = httpClientFactory.CreateClient(); var response = await client.PostAsJsonAsync($"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={WebUtility.UrlEncode(key)}", payload); using var json = JsonDocument.Parse(await response.Content.ReadAsStringAsync()); var text = json.RootElement.GetProperty("candidates")[0].GetProperty("content").GetProperty("parts")[0].GetProperty("text").GetString() ?? "Gemini returned no readable text."; return new() { ["reply"] = text, ["configured"] = true }; }
        catch { return new() { ["reply"] = "Gemini is configured but unavailable. Verify the key, model access, and network connection.", ["configured"] = true, ["providerError"] = true }; }
    }

    public async Task<bool> SendEmailAsync(EmailRequest request)
    {
        var from = configuration["Mail:Username"] ?? Environment.GetEnvironmentVariable("MAIL_USERNAME"); if (string.IsNullOrWhiteSpace(from) || string.IsNullOrWhiteSpace(request.To)) return false;
        using var client = new SmtpClient(configuration["Mail:Host"] ?? Environment.GetEnvironmentVariable("MAIL_HOST") ?? "localhost", int.TryParse(configuration["Mail:Port"] ?? Environment.GetEnvironmentVariable("MAIL_PORT"), out var port) ? port : 587) { EnableSsl = true, Credentials = new NetworkCredential(from, configuration["Mail:Password"] ?? Environment.GetEnvironmentVariable("MAIL_PASSWORD")) };
        await client.SendMailAsync(new MailMessage(from, request.To, request.Subject ?? "AssetFlow notification", request.Body ?? "")); return true;
    }

    private async Task<long> CountAsync(string sql, IReadOnlyDictionary<string, string?> c, bool employee) { await using var db = Connection(); await db.OpenAsync(); await using var cmd = new MySqlCommand(sql, db); cmd.Parameters.AddWithValue("@company", Company(c)); if (employee) cmd.Parameters.AddWithValue("@user", User(c)); return Convert.ToInt64(await cmd.ExecuteScalarAsync()); }
    private async Task<int> ExecuteAsync(string sql, params (string Name, object? Value)[] values) { await using var db = Connection(); await db.OpenAsync(); await using var cmd = new MySqlCommand(sql, db); foreach (var v in values) cmd.Parameters.AddWithValue(v.Name, v.Value ?? DBNull.Value); return await cmd.ExecuteNonQueryAsync(); }
    private static async Task<List<Dictionary<string, object?>>> QueryAsync(MySqlConnection db, string sql, params (string Name, object? Value)[] values) { await using var cmd = new MySqlCommand(sql, db); foreach (var v in values) cmd.Parameters.AddWithValue(v.Name, v.Value ?? DBNull.Value); await using var reader = await cmd.ExecuteReaderAsync(); var rows = new List<Dictionary<string, object?>>(); while (await reader.ReadAsync()) { var row = new Dictionary<string, object?>(); for (var i = 0; i < reader.FieldCount; i++) row[reader.GetName(i)] = await reader.IsDBNullAsync(i) ? null : reader.GetValue(i); rows.Add(row); } return rows; }
}

File: backend/assetflow-notification-service/src/main/java/com/assetflow/notification/service/NotificationDataService.java
package com.assetflow.notification.service;

import java.util.*;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestClient;

@Service
public class NotificationDataService {
  private final JdbcTemplate db;
  private final String geminiKey;
  private final String geminiModel;

  public NotificationDataService(JdbcTemplate d,
                                 @Value("${app.gemini.api-key:}") String k,
                                 @Value("${app.gemini.model:gemini-2.0-flash}") String m) {
    this.db = d;
    this.geminiKey = k;
    this.geminiModel = m;
  }

  private long number(Object value) {
    if (value instanceof Number n) return n.longValue();
    return Long.parseLong(String.valueOf(value));
  }

  private long company(Map<String,Object> c) {
    return number(c.get("companyId"));
  }

  private long user(Map<String,Object> c) {
    return number(c.get("sub"));
  }

  public List<Map<String,Object>> notifications(Map<String,Object> c) {
    return db.queryForList("select * from notifications where company_id=? and (user_id=? or user_id=0) order by created_at desc", company(c), user(c));
  }

  @Transactional
  public Map<String,Object> createNotification(Map<String,Object> c, Map<String,Object> b) {
    long userId = b.get("userId") != null ? number(b.get("userId")) : (b.get("user_id") != null ? number(b.get("user_id")) : user(c));
    String title = String.valueOf(b.getOrDefault("title", "Notification"));
    String message = String.valueOf(b.getOrDefault("message", ""));
    db.update("insert into notifications(company_id,user_id,title,message) values(?,?,?,?)",
        company(c), userId, title, message);
    Long notifId = db.queryForObject("select notification_id from notifications where company_id=? and user_id=? order by notification_id desc limit 1", Long.class, company(c), userId);
    Map<String, Object> res = new LinkedHashMap<>(b);
    res.put("notification_id", notifId);
    res.put("id", notifId);
    res.put("is_read", false);
    return res;
  }

  public void read(Map<String,Object> c, long id) {
    db.update("update notifications set is_read=true where notification_id=? and company_id=?", id, company(c));
  }

  public void delete(Map<String,Object> c, long id) {
    db.update("delete from notifications where notification_id=? and company_id=? and (user_id=? or user_id=0)", id, company(c), user(c));
  }

  public void readAll(Map<String,Object> c) {
    db.update("update notifications set is_read=true where company_id=? and (user_id=? or user_id=0)", company(c), user(c));
  }

  public List<Map<String,Object>> audits(Map<String,Object> c) {
    String sql = "select a.*, a.module as entity_name, concat(u.first_name, ' ', u.last_name) as user_name, u.email as user_email from audit_logs a left join users u on u.user_id=a.user_id where a.company_id=?";
    if ("EMPLOYEE".equals(String.valueOf(c.get("role")))) {
      return db.queryForList(sql + " and a.user_id=? order by a.created_at desc limit 200", company(c), user(c));
    }
    return db.queryForList(sql + " order by a.created_at desc limit 200", company(c));
  }

  @Transactional
  public Map<String,Object> audit(Map<String,Object> c, Map<String,Object> b) {
    String module = String.valueOf(b.getOrDefault("module", b.getOrDefault("moduleName", b.getOrDefault("entityName", b.getOrDefault("entity_name", "GENERAL")))));
    String action = String.valueOf(b.getOrDefault("action", "AUDIT_EVENT"));
    String entityId = b.get("entityId") != null ? String.valueOf(b.get("entityId")) : (b.get("entity_id") != null ? String.valueOf(b.get("entity_id")) : null);
    String description = b.get("description") != null ? String.valueOf(b.get("description")) : (b.get("details") != null ? String.valueOf(b.get("details")) : "Audit record");
    String ipAddress = b.get("ipAddress") != null ? String.valueOf(b.get("ipAddress")) : (b.get("ip_address") != null ? String.valueOf(b.get("ip_address")) : "127.0.0.1");
    db.update("insert into audit_logs(company_id,user_id,module,action,entity_id,description,ip_address) values(?,?,?,?,?,?,?)",
        company(c), user(c), module, action, entityId, description, ipAddress);
    Map<String, Object> res = new LinkedHashMap<>(b);
    res.put("module", module);
    res.put("entity_name", module);
    return res;
  }

  public Map<String,Object> dashboard(Map<String,Object> c) {
    Map<String,Object> d = new LinkedHashMap<>();
    long tenant = company(c);
    boolean employee = "EMPLOYEE".equals(String.valueOf(c.get("role")));
    String scope = employee ? " and user_id=?" : "";
    Object[] notificationArgs = employee ? new Object[]{tenant, user(c)} : new Object[]{tenant};
    d.put("totalAssets", safeCount("select count(*) from assets where company_id=? and is_active=true", tenant));
    d.put("total_assets", d.get("totalAssets"));
    d.put("allocatedAssets", safeCount("select count(*) from assets where company_id=? and is_active=true and status='ALLOCATED'", tenant));
    d.put("availableAssets", safeCount("select count(*) from assets where company_id=? and is_active=true and status='AVAILABLE'", tenant));
    d.put("maintenanceCount", safeCount("select count(*) from maintenance where company_id=? and is_active=true and status not in ('COMPLETED','CANCELLED')", tenant));
    d.put("notifications", db.queryForObject("select count(*) from notifications where company_id=?" + scope, Long.class, notificationArgs));
    d.put("unreadNotifications", db.queryForObject("select count(*) from notifications where company_id=? and is_read=false" + scope, Long.class, notificationArgs));
    d.put("auditEvents", db.queryForObject("select count(*) from audit_logs where company_id=?" + (employee ? " and user_id=?" : ""), Long.class, employee ? new Object[]{tenant, user(c)} : new Object[]{tenant}));
    return d;
  }

  public List<Map<String,Object>> report(String type, Map<String,Object> c) {
    String sql = switch (type) {
      case "assets" -> "select status, count(*) as total from assets where company_id=? and is_active=true group by status";
      case "users", "employees" -> "select r.role_name, count(*) as total from users u join roles r on r.role_id=u.role_id where u.company_id=? and u.is_active=true group by r.role_name";
      case "departments" -> "select d.department_name, count(u.user_id) as total from departments d left join users u on u.department_id=d.department_id and u.company_id=d.company_id and u.is_active=true where d.company_id=? and d.is_active=true group by d.department_id, d.department_name";
      case "vendors" -> "select is_active, count(*) as total from vendors where company_id=? group by is_active";
      case "maintenance" -> "select status, count(*) as total from maintenance where company_id=? and is_active=true group by status";
      case "requests" -> "select status, count(*) as total from asset_requests where company_id=? and is_active=true group by status";
      case "audit" -> "select module, count(*) as total from audit_logs where company_id=? group by module order by total desc";
      default -> "select status, count(*) as total from purchase_orders where company_id=? and is_active=true group by status";
    };
    return db.queryForList(sql, company(c));
  }

  public Map<String,Object> chat(Map<String,Object> body) {
    String prompt = String.valueOf(body.getOrDefault("message", body.getOrDefault("prompt", body.getOrDefault("query", ""))));
    if (geminiKey == null || geminiKey.isBlank()) {
      return Map.of("reply", "Gemini is not configured yet. Set GEMINI_API_KEY in the notification service environment to enable company-aware answers.", "configured", false);
    }
    Map<String,Object> payload = Map.of("contents", List.of(Map.of("parts", List.of(Map.of("text", prompt)))));
    try {
      Map<?,?> result = RestClient.create().post().uri("https://generativelanguage.googleapis.com/v1beta/models/" + geminiModel + ":generateContent?key=" + geminiKey).body(payload).retrieve().body(Map.class);
      String text = extractText(result);
      return Map.of("reply", text, "configured", true);
    } catch (Exception ex) {
      return Map.of("reply", "Gemini is configured but unavailable. Verify the key, model access, and network connection.", "configured", true, "providerError", true);
    }
  }

  private String extractText(Map<?,?> response) {
    try {
      Object candidates = response.get("candidates");
      if (candidates instanceof List<?> list && !list.isEmpty()) {
        Object content = ((Map<?,?>) list.get(0)).get("content");
        Object parts = ((Map<?,?>) content).get("parts");
        if (parts instanceof List<?> p && !p.isEmpty()) {
          return String.valueOf(((Map<?,?>) p.get(0)).get("text"));
        }
      }
    } catch (Exception ignored) {}
    return "Gemini returned no readable text.";
  }

  public Map<String,Object> chat(Map<String,Object> body, Map<String,Object> claims) {
    String question = String.valueOf(body.getOrDefault("message", body.getOrDefault("prompt", body.getOrDefault("query", ""))));
    if (geminiKey == null || geminiKey.isBlank()) {
      Map<String,Object> fallback = Map.of("reply", "Gemini is not configured yet. Set GEMINI_API_KEY in the notification service environment to enable company-aware answers.", "configured", false);
      try {
        db.update("insert into ai_chat_history(company_id,user_id,question,answer,message,sender_type) values(?,?,?,?,?,?)", company(claims), user(claims), question, fallback.get("reply"), question, "USER");
      } catch (Exception ignored) {}
      return fallback;
    }
    long tenant = company(claims);
    Map<String,Object> context = new LinkedHashMap<>();
    context.put("assets", safeCount("select count(*) from assets where company_id=? and is_active=true", tenant));
    context.put("employees", safeCount("select count(*) from users where company_id=? and is_active=true", tenant));
    context.put("departments", safeCount("select count(*) from departments where company_id=? and is_active=true", tenant));
    context.put("openRequests", safeCount("select count(*) from asset_requests where company_id=? and is_active=true and status in ('PENDING','OPEN')", tenant));
    context.put("openMaintenance", safeCount("select count(*) from maintenance where company_id=? and is_active=true and status not in ('COMPLETED','CANCELLED')", tenant));
    String grounded = "You are AssetFlow's internal assistant. Answer only from the tenant context below and general asset-management guidance. Never invent tenant records, IDs, users, or cross-company data. If the context is insufficient, say so. Tenant context: " + context + ". User question: " + question;
    Map<String,Object> result = chat(Map.of("message", grounded));
    try {
      db.update("insert into ai_chat_history(company_id,user_id,question,answer,message,sender_type) values(?,?,?,?,?,?)", tenant, user(claims), question, result.get("reply"), question, "USER");
    } catch (Exception ignored) {}
    return result;
  }

  private long safeCount(String sql, long tenant) {
    try {
      Long value = db.queryForObject(sql, Long.class, tenant);
      return value == null ? 0 : value;
    } catch (Exception ignored) {
      return 0;
    }
  }
}
